File: mappers-lua/mmc3.lua
MapperName = "MMC3"
MapperNumber = 4
DefaultPrgSize = 512 * 1024
DefaultChrSize = 256 * 1024

function DumpPrg(size)
	local banks = size / 0x2000
	for b = 0, banks-3, 2 do
		print("Reading PRG banks #" .. tostring(b) .. " and #" .. tostring(b+1) .. "...")
		WriteCpu(0x8000, {6, b})
		WriteCpu(0x8000, {7, b+1})
		ReadAddPrg(0x8000, 0x4000)
	end

	print("Reading last PRG banks #" .. tostring(banks-2) .. " and #" .. tostring(banks-1) .. "...")
	ReadAddPrg(0xC000, 0x4000)
end

function DumpChr(size)
	local banks = size / 0x0400
	for b = 0, banks-1, 8 do
		print("Reading CHR banks #" .. tostring(b) .. ", #" .. tostring(b+1) .. ", #" .. tostring(b+2) .. ", #" .. tostring(b+3) .. ", #" .. tostring(b+4) .. ", #" .. tostring(b+5) .. ", #" .. tostring(b+6) .. " and #" .. tostring(b+7) .. "...")
		WriteCpu(0x8000, {0, b})
		WriteCpu(0x8000, {1, b+2})
		WriteCpu(0x8000, {2, b+4})
		WriteCpu(0x8000, {3, b+5})
		WriteCpu(0x8000, {4, b+6})
		WriteCpu(0x8000, {5, b+7})
		ReadAddChr(0x0000, 0x2000)
	end
end

function EnablePrgRam(size)
	WriteCpu(0xA001, {0x80})
end

File: mappers-cs/MMC3.cs
using System;
using System.Collections.Generic;

namespace com.clusterrr.Famicom.Mappers
{
    public class MMC3 : IMapper
    {
        public string Name
        {
            get { return "MMC3"; }
        }

        public int Number
        {
            get { return 4; }
        }

        public string UnifName
        {
            get { return null; }
        }

        public int DefaultPrgSize
        {
            get { return 512 * 1024; }
        }

        public int DefaultChrSize
        {
            get { return 256 * 1024; }
        }

        public void DumpPrg(FamicomDumperConnection dumper, List<byte> data, int size)
        {
            byte banks = (byte)(size / 0x2000);
            for (byte bank = 0; bank < banks - 2; bank += 2)
            {
                Console.Write("Reading PRG banks #{0} and #{1}... ", bank, bank + 1);
                dumper.WriteCpu(0x8000, new byte[] { 6, bank });
                dumper.WriteCpu(0x8000, new byte[] { 7, (byte)(bank | 1) });
                data.AddRange(dumper.ReadCpu(0x8000, 0x4000));
                Console.WriteLine("OK");
            }
            Console.Write("Reading last PRG banks #{0} and #{1}... ", banks - 2, banks - 1);
            data.AddRange(dumper.ReadCpu(0xC000, 0x4000));
            Console.WriteLine("OK");
        }

        public void DumpChr(FamicomDumperConnection dumper, List<byte> data, int size)
        {
            int banks = size / 0x400;
            if (banks > 256) throw new Exception("CHR size is too big");
            for (int bank = 0; bank < banks; bank += 4)
            {
                Console.Write("Reading CHR banks #{0}, #{1}, #{2}, #{3}... ", bank, bank + 1, bank + 2, bank + 3);
                dumper.WriteCpu(0x8000, new byte[] { 2, (byte)bank });
                dumper.WriteCpu(0x8000, new byte[] { 3, (byte)(bank | 1) });
                dumper.WriteCpu(0x8000, new byte[] { 4, (byte)(bank | 2) });
                dumper.WriteCpu(0x8000, new byte[] { 5, (byte)(bank | 3) });
                data.AddRange(dumper.ReadPpu(0x1000, 0x1000));
                Console.WriteLine("OK");
            }
        }

        public void EnablePrgRam(FamicomDumperConnection dumper)
        {
            dumper.WriteCpu(0xA001, 0x80);
        }
    }
}
